### PR TITLE
 (#2) Replacing SLF4J with Log4J 2 LogBuilder

### PR DESCRIPTION
**Pull request by MatthewFoster624**
First pass at cleaning up the logging & messages. 
Added consistent formatting & detail within log messages.
Uses the Log4J 2 (2.13.0) LogBuilder format.

Note - Did not edit the following portions of the code:
syncthing-client-cli,
synthing-repository-default, and
syncthing-temp-repository-encryption.

There are 2 TODOs related to messaging in this set of code, both in `KeystoreHandler.kt`

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -16,6 +17,7 @@ buildscript {
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
         classpath 'com.github.ben-manes:gradle-versions-plugin:0.17.0'
         classpath 'com.google.protobuf:protobuf-gradle-plugin:0.8.11'
+        classpath "org.apache.logging.log4j:log4j-api:$log4JVersion"
```
Comment (MariusVolkhart) [inline build.gradle]: Not here. `classpath` refers to the gradle system itself, not the code you're compiling.

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,6 +1,6 @@
-#Fri Sep 14 08:50:38 CEST 2018
+#Sat Feb 08 18:08:13 EST 2020
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.1.1-all.zip
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.1.1-bin.zip
+zipStoreBase=GRADLE_USER_HOME
```
Comment (MariusVolkhart) [inline gradle/wrapper/gradle-wrapper.properties]: undo changes in this file please. We don't want `all`, we want `bin`

File: settings.gradle [hunk 3]
```diff
@@ -1 +1,2 @@
 include ':syncthing-repository-default', ':syncthing-relay-client', ':syncthing-bep', ':syncthing-core', ':syncthing-client', ':syncthing-discovery', ':syncthing-client-cli', ':syncthing-temp-repository-encryption'
+
```
Comment (MariusVolkhart) [inline settings.gradle]: unnecessary whitespace change

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 4]
```diff
@@ -32,7 +34,7 @@ import java.util.*
 import kotlin.collections.HashMap
 
 object BlockPuller {
-    private val logger = LoggerFactory.getLogger(javaClass)
+    private val LOGGER = LogManager.getLogger(javaClass)
```
Comment (MariusVolkhart) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: use `BlockPuller::class.java` instead of `javaClass`

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 5]
```diff
@@ -51,10 +53,12 @@ object BlockPuller {
         val (newFileInfo, fileBlocks) = indexHandler.getFileInfoAndBlocksByPath(fileInfo.folder, fileInfo.path) ?: throw FileNotFoundException()
 
         if (fileBlocks.hash != fileInfo.hash) {
-            throw IllegalStateException("the current file entry hash does not match the hash of the provided one")
+            val exception = IllegalStateException("The current file entry hash does not match the hash of the provided one.")
+            LOGGER.atInfo().withThrowable(exception).log("Expected hash: ( {} ), Provided hash: ( {} ).",fileInfo.hash,fileBlocks.hash)
```
Comment (MariusVolkhart) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: Don't include the exception, since it is being generated at this location. This should also be at least `warn` level. Use a message like `The current file entry hash ({}) does not match the provided one ({}).`. No spaces inside the parenthesis. In this case, since the exception is going to get thrown anyhow, I'd write it like,

```kotlin
val message = "The current file entry hash (${fileInfo.hash}) does not match the provided one (${fileBlocks.hash})."
val exception = IllegalStateException(message)
LOGGER.atError().log(message)
throw exception
```

Make sure to put spaces after your method argument commas.

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 6]
```diff
@@ -51,10 +53,12 @@ object BlockPuller {
...
+            throw exception
         }
 
-        logger.info("pulling file = {}", fileBlocks)
+        LOGGER.atInfo().log("Pulling File: {}, File Blocks: {}.", fileInfo.path, fileBlocks)
```
Comment (MariusVolkhart) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: Look at the `toString()` method of `fileBlocks`. It already includes a `path`. Is that the same path as `fileInfo.path`?

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 7]
```diff
@@ -65,7 +69,7 @@ object BlockPuller {
         )
 
         suspend fun pullBlock(fileBlocks: FileBlocks, block: BlockInfo, timeoutInMillis: Long, connectionActorWrapper: ConnectionActorWrapper): ByteArray {
-            logger.debug("sent message for block, hash = {}", block.hash)
+            LOGGER.atDebug().log("Request sent for the block with this hash: {}.", block.hash)
```
Comment (MariusVolkhart) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: "Request sent for block {}."

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 8]
```diff
@@ -84,22 +88,27 @@ object BlockPuller {
                             // is handled differently so that the timeout is ignored.
                             // Due to that, it's converted to an IOException.
 
-                            throw IOException("timeout during requesting block")
+                            val exception = IOException("Time limit exceeded while requesting block.")
+                            LOGGER.atInfo().withThrowable(exception).log("Timeout set for {} seconds.", box(timeoutInMillis/1000))
```
Comment (MariusVolkhart) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: careful with things like `timeoutInMillis/1000`. You're going to lose accuracy, because the math is done entirely in whole numbers. You need to explicitly ask for decimals if you want them. In this case, just log it as millis. Using seconds isn't a big benefit.

spaces around operators like `/`

Again, if we're throwing an exception, it's an `error` in the logs

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 9]
```diff
@@ -84,22 +88,27 @@ object BlockPuller {
...
                         }
                     }
 
             if (response.code != BlockExchangeProtos.ErrorCode.NO_ERROR) {
                 // the server does not have/ want to provide this file -> don't ask him again
                 connectionHelper.disableConnection(connectionActorWrapper)
-
-                throw IOException("received error response ${response.code}")
+                val exception = IOException("Server does not have or is not able to provide the requested file.")
+                LOGGER.atDebug().withThrowable(exception).log("Received following error response from server: {}.", response.code)
```
Comment (MariusVolkhart) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: `box()`? Don't include the exception. error level. log & exception should have similar information in their message

**Review by MariusVolkhart — CHANGES_REQUESTED** [review]

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt (hunk 6 above)
Comment (MatthewFoster624) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: I think so. This has been an adventure into Kotlin for me... but I think -- from tracing this back thru the `getFileInfoAndBlocksByPath` function and the Class definitions of `FileBlocks.kt` & `FileInfo.kt` -- that there is some duplicate information being passed in the `Pair.of(fileInfo, fileBlocks)`.

Comment (MatthewFoster624) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: Is this correctly interpreting the Kotlin function?
`fun getFileInfoAndBlocksByPath(folder: String, path: String): Pair<FileInfo, FileBlocks>? {`
means:
Received arguments: `folder`, `path` (both strings)
`:` --> means Returns
Returned arguments: `<FileInfo, FileBlocks>` (as pair, defined by their respective class definitions)
`?` --> means can return `null`

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt (hunk 9 above)
Comment (MatthewFoster624) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: It is not a boxable argument as far as I understand.

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/connectionactor/HelloMessageHandler.kt [hunk 10]
```diff
@@ -24,6 +24,7 @@ import java.io.DataInputStream
 import java.io.DataOutputStream
 import java.io.IOException
 
+//TODO This Logger is named differently. Why? Make Consistent?
```
Comment (MatthewFoster624) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/connectionactor/HelloMessageHandler.kt]: This TODO I'd like to address in this commit. Comment describes my questions.

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt (hunk 6 above)
Comment (MariusVolkhart) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: yes, that's correct. nicely done

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/connectionactor/HelloMessageHandler.kt (hunk 10 above)
Comment (MariusVolkhart) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/connectionactor/HelloMessageHandler.kt]: I'm not sure why. yes, make it consistent

Comment (MatthewFoster624) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/connectionactor/HelloMessageHandler.kt]: Alright, I need to talk through this one with you.
This one is unique in that it's not contained in a `class { ...}` or `object { ... }` block.
`private val LOGGER = LogManager.getLogger(HelloMessageHandler::Class.java)` triggers an error where `HelloMessageHandler` is an unresolved reference. Not sure if I need to make this whole script a class, or how to ask the right question here.

**Comment by MatthewFoster624** [comment]
Alright, if you agree, I'll squash & rebase.

**Review by MariusVolkhart — APPROVED** [review]
